spicy/toolchain/include/compiler/detail/codegen/productions/while.h

```diff
@@ -8,33 +8,70 @@
 
 #include <spicy/ast/types/unit.h>
 #include <spicy/compiler/detail/codegen/production.h>
+#include <spicy/compiler/detail/codegen/productions/look-ahead.h>
 
 namespace spicy::detail::codegen::production {
 
 /**
- * A production executing as long as a given boolean expression evaluates
- * to true.
+ * A production executing as long as either a given boolean expression evaluates
+ * to true, or, if no expression is provided, as determined by look-ahead symbols
  */
 class While : public ProductionBase, public spicy::trait::isNonTerminal {
 public:
+    /**
+     * Constructor for a while-loop using an expression as the condition for termination.
+     */
     While(const std::string& symbol, Expression e, Production body, const Location& l = location::None)
-        : ProductionBase(symbol, l), _expression(std::move(e)), _body(std::move(body)) {}
+        : ProductionBase(symbol, l), _body(std::move(body)), _expression(std::move(e)) {}
 
-    const Expression& expression() const { return _expression; }
-    const Production& body() const { return _body; }
+    /**
+     * Constructor for a while-loop using look-ahead as the condition for
+     * termination. When using this constructor, `preprocessLookAhead()` must
```

Comment:
Could we instead pass the grammar into this constructor to make it easier to guarantee that exactly on of `_expression` or `_body_for_grammar` is set?

Reply:
AST classes should be instantiable without having a grammar (yet), that's why this is a little bit awkward.